### DISCUSSION TITLE
Klopfenstein taper implementation

### DISCUSSION THREAD

**Discussion by ZiadHatab** [scikit-rf/scikit-rf#1059] [Ideas]
Hi!

I'm trying to contribute to the Taper module, but I'm not sure I understood the purpose behind the module. Is it about tapering the characteristic impedance based on an impedance profile, or is it just about tapering any parameter in a media class following some function?

What confuses me is that in the Taper class there's a placeholder for an unimplemented function for a [Klopfenstein taper](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.taper.Klopfenstein.html). I thought I'd help out and implement the taper, as I've already written code for the Klopfenstein taper (https://github.com/ZiadHatab/klopf-taper). But I'm not sure how to fit this in with the taper class currently implemented, as the Klopfenstein taper is by definition only a taper from Z1 to Z2. We'd need to reverse calculate the parameter for the desired transmission line.

For example, if I look at the microstrip [example](https://scikit-rf.readthedocs.io/en/latest/examples/taper/tapered_transmission_lines.html) where the width of the strip is tapered as a linear function between w1=20 mil and w2=90 mil, and I plot the characteristic impedance as a function of width, I wouldn't get a linear response for the characteristic impedance, regardless of the frequency I'm looking at. Obviously, this is because the characteristic impedance of a microstrip line is not linear with respect to width.
![image](https://github.com/scikit-rf/scikit-rf/assets/46247580/800fb7aa-f4cf-4e1b-a19a-2e050da7f77d)
```python
import numpy as np
import matplotlib.pyplot as plt
import skrf as rf
from skrf.media import MLine

if __name__ == '__main__':
    # Model Parameters
    freq = rf.Frequency(1, 20, unit='GHz', npoints=20)
    w1 = 20*rf.mil     # conductor width [m]
    w2 = 90*rf.mil     # conductor width [m]
    h = 20*rf.mil      # dielectric thickness [m]
    t = 0.7*rf.mil     # conductor thickness [m]
    rho = 1.724138e-8  # Copper resistivity [Ohm.m]
    ep_r = 10          # dielectric relative permittivity
    rough = 1e-6       # conductor RMS roughtness [m]
    taper_length = 200*rf.mil  # [m]
    n_sections = 50 # number of segments between w1 and w2
    
    # create a 2-port Network
    taper_linear = rf.taper.Linear(med=MLine, param='w', start=w1, stop=w2,
                        length=taper_length, n_sections=n_sections,
                        med_kw={'frequency': freq, 'h': h, 't':t, 'ep_r': ep_r,
                                'rough': rough, 'rho': rho}).network
    
    Z0 = []
    ws = np.linspace(w1, w2, n_sections)
    for w in ws:
        mline = MLine(w=w, frequency=freq, h=h, t=t, ep_r=ep_r, rough=rough, rho=rho)
        Z0.append(mline.z0)
    Z0 = np.array(Z0).T

    plt.figure()
    for z,f in zip(Z0,freq.f):
        plt.plot(ws/rf.mil, z.real, label=f'@f = {f*1e-9:.0f} GHz')
    plt.xlabel('Width (mil)')
    plt.ylabel('Impedance real part (Ohm)')
    plt.legend(ncol=3)
    plt.ylim([10, 70])

    plt.show()
````

So, I think the taper implementation as is would make sense if the Media class had a synthesis capability to reverse calculate a parameter from a given impedance value at a certain frequency. For instance, in the microstrip example above, the taper function should have computed Z1 from w1 and Z2 from w2, then created N segments as a linear function between Z1 and Z2 and mapped every Z value to a w value through reverse calculation. The reverse calculation could be done as a black box using the root solver from the [SciPy optimization library](https://docs.scipy.org/doc/scipy/reference/optimize.html).

I'd love to help with implementing the Klopfenstein taper, but I'd appreciate feedback on how to do it with the current structure of the taper module. Thanks!

**Comment by arsenovic** [comment]
from the taper docs, 

```
 Note
----
The default behaviour should is to taper based on impedance.
To do this we inspect the `med` class for a `from_z0`
init method, and if it exists, we assign it to `med` attribute,
in `__init__`.
Admittedly having `med` be a class or a method is abuse,
it makes for a intuitive operation.```
```
does that help you?

this would be a great addition.

**Comment by ZiadHatab** [comment]
OK. Understood. 
I haven't touched this in a while. I will look into it.
